internal/engine/upper.go

```diff
@@ -224,6 +224,94 @@ func handleBuildStarted(ctx context.Context, state *store.EngineState, action bu
 	removeFromTriggerQueue(state, mn)
 }
 
+func updateBuildStatus(status *store.BuildStatus, br model.BuildRecord, result store.BuildResult, isBuildSuccess bool) {
+	// Record the last result
+	status.LastResult = result
+
+	// Remove pending file changes that were consumed by this build.
+	for file, modTime := range status.PendingFileChanges {
+		if store.BeforeOrEqual(modTime, br.StartTime) {
+			delete(status.PendingFileChanges, file)
+		}
+	}
+
+	// Remove pending dependency changes that were consumed by this build.
+	for id, modTime := range status.PendingDependencyChanges {
+		if store.BeforeOrEqual(modTime, br.StartTime) {
+			delete(status.PendingDependencyChanges, id)
+		}
+	}
+
+	// If the whole build succeeded, mark all of the images in the
+	// build as successful.
+	//
+	// This has a minor bug: if one of the images in a set failed, but another one
+	// succeeded, this doesn't mark any of them as successful. We should fix this,
+	// though personally, I think LastSuccessfulResult has problematic semantics
+	// to begin with, and it might make sense to replace it entirely.
+	if isBuildSuccess {
+		status.LastSuccessfulResult = result
+	}
+}
+
+// When a Manifest build finishes, update the BuildStatus for all applicable
+// targets in the engine state.
+func handleBuildResults(engineState *store.EngineState,
+	mt *store.ManifestTarget, br model.BuildRecord, results store.BuildResultSet) {
+	isBuildSuccess := br.Error == nil
+
+	// Update all the build statuses for the current manifest.
+	for _, target := range mt.Manifest.TargetSpecs() {
+		status := mt.State.MutableBuildStatus(target.ID())
+		updateBuildStatus(status, br, results[target.ID()], isBuildSuccess)
+	}
+
+	// Update build statuses for duplicated image targets in other manifests.
+	// This ensures that those images targets aren't rebuilt.
+	engineState.ForEachMutableBuildStatusInResults(results, func(id model.TargetID, result store.BuildResult, mn model.ManifestName, status *store.BuildStatus) {
+		if mt.Manifest.Name == mn {
+			return
+		}
+
+		_, isImageResult := result.(store.ImageBuildResult)
+		if !isImageResult {
+			return
+		}
+
+		updateBuildStatus(status, br, result, isBuildSuccess)
+	})
+
+	// Suppose we built manifestA, which contains imageA depending on imageCommon.
+	//
+	// We also have manifestB, which contains imageB depending on imageCommon.
+	//
+	// We need to mark imageB as dirty, because it was not built in the manifestA
+	// build but its dependency was built.
+	engineState.ForEachMutableBuildStatusDependingOn(results, func(id, reverseDepID model.TargetID, mn model.ManifestName, reverseDepStatus *store.BuildStatus) {
```

Comment:
the naming scheme `id` and `reverseDepID` had me tying myself in knots, is there a more straightforward naming scheme? (like say `parent` and `child`?) Alternately, could clarify in re: the example above--
> In terms of the above example: `id` refers to `imageCommon`, `reverseDepId` refers to `imageB`

Reply:
if it makes you feel any better: the argument about what to name this relationship comes up on every build project.

"reverse deps" is kind of the term of art in this space, see:
https://docs.bazel.build/versions/master/query.html#rdeps
because even though it's awkward, it's at least unambiguous